File: 2024/AoC2024/Sources/Tasks/Day2.swift
import Foundation
internal import Algorithms
import Framework

public final class Day2: Day<Task, Int> {
    public override func perform(data: String, param: P) throws -> R {
        let reports = try parse(data)

        return switch param {
        case .task1: task1(reports: reports)
        case .task2: task2(reports: reports)
        }
    }
    
    private func parse(_ data: String) throws -> ([[Int]]) {
        try data.lines.map { line in
            try line.words.map(toInt)
        }
    }
    
    private func task1(reports: [[Int]]) -> R {
        reports.count(where: checkReport)
    }
    
    private func task2(reports: [[Int]]) -> R {
        reports.count { report in
            if checkReport(report) {
                return true
            }
            
            for i in 0..<report.count {
                var reportCopy = report
                reportCopy.remove(at: i)
                if checkReport(reportCopy) {
                    return true
                }
            }
            return false
        }
    }
    
    private func checkReport(_ report: [Int]) -> Bool {
        let diffs = report.adjacentPairs().map({ $0.0 - $0.1})
        return diffs.allSatisfy({ abs($0) >= 1 && abs($0) <= 3}) && diffs.dropFirst().allSatisfy {
            guard let firstDiff = diffs.first else {
                return false
            }
            return $0.signum() == firstDiff.signum()
        }
    }
}

File: typescript/2021/src/tasks/__tests__/day2.ts
import { realDataPath, test, testDataPath } from "../../testUtil";
import { task1, task2 } from "../day2";

describe("task 1", () => {
  it("test data", test(task1, testDataPath("day2"), 150));
  it("real data", test(task1, realDataPath("day2"), 1524750));
});

describe("task 2", () => {
  it("test data", test(task2, testDataPath("day2"), 900));
  it("real data", test(task2, realDataPath("day2"), 1592426537));
});
